### ISSUE TITLE
Reload css only by injecting javascript

### ISSUE THREAD

**Issue by nicola** [mozilla/fireplay-sublime#1]
I just found out that the way we can reload CSS can be very simple.
Instead of going through the debugger and therefore hope that we mapped the CSS correctly. We can instead inject and execute a js script from the console every time we save a file which is `.css`, `.styl`, `.less`, `.sass`..

The script looks like this:

```
CSSStyleSheet.prototype.reload = function reload(){
  // Reload one stylesheet
  // usage: document.styleSheets[0].reload()
  // return: URI of stylesheet if it could be reloaded, overwise undefined
  if (this.href) {
    var href = this.href;
    var i = href.indexOf('?'),
        last_reload = 'last_reload=' + (new Date).getTime();
    if (i < 0) {
      href += '?' + last_reload;
    } else if (href.indexOf('last_reload=', i) < 0) {
      href += '&' + last_reload;
    } else {
      href = href.replace(/last_reload=\d+/, last_reload);
    }
    return this.ownerNode.href = href;
  }
};

StyleSheetList.prototype.reload = function reload(){
  // Reload all stylesheets
  // usage: document.styleSheets.reload()
  for (var i=0; i<this.length; i++) {
    this[i].reload()
  }
};
```

and therefore just calling `document.styleSheets.reload()`

**Comment by nicola** [comment]
However I am sure that through the debugger we could try to reload all the css nodes without passing through javascript.

The new approach of reloading all the css when editing a css file still holds.

**Comment by nicola** [comment]
Potentially this can be done through RDP:

```
DBG-SERVER: Packet 545 sent to "conn0.console20"
DBG-SERVER: Received packet 545: {
  "to": "conn0.console20",
  "type": "evaluateJS",
  "text": "document.styleSheets[0].ownerNode.href = \"http://nicolagreco.com/test/test.css\"",
  "frameActor": null
}
DBG-SERVER: Packet 546 sent from "conn0.console20"
DBG-SERVER: Packet 547 sent from "conn0.domwalker39"
DBG-SERVER: Received packet 546: {
  "from": "conn0.console20",
  "input": "document.styleSheets[0].ownerNode.href = \"http://nicolagreco.com/test/test.css\"",
  "result": "http://nicolagreco.com/test/test.css",
  "timestamp": 1405720367528,
  "exception": null,
  "helperResult": null
}
DBG-SERVER: Received packet 547: {
  "type": "newMutations",
  "from": "conn0.domwalker39"
}
DBG-SERVER: Packet 548 sent to "conn0.domwalker39"
DBG-SERVER: Received packet 548: {
  "type": "getMutations",
  "cleanup": true,
  "to": "conn0.domwalker39"
}
DBG-SERVER: Packet 549 sent from "conn0.domwalker39"
DBG-SERVER: Received packet 549: {
  "mutations": [
    {
      "type": "attributes",
      "target": "conn0.domnode55",
      "attributeName": "href",
      "newValue": "http://nicolagreco.com/test/test.css"
    }
  ],
  "from": "conn0.domwalker39"
}
DBG-SERVER: Packet 550 sent to "conn0.domwalker39"
DBG-SERVER: Received packet 550: {
  "type": "children",
  "node": "conn0.domnode43",
  "maxNodes": 10,
  "whatToShow": 1,
  "to": "conn0.domwalker39"
}
DBG-SERVER: Packet 551 sent from "conn0.domwalker39"
DBG-SERVER: Received packet 551: {
  "hasFirst": true,
  "hasLast": true,
  "nodes": [
    {
      "actor": "conn0.domnode44",
      "baseURI": "http://nicolagreco.com/test/",
      "parent": "conn0.domnode43",
      "nodeType": 1,
      "namespaceURI": "http://www.w3.org/1999/xhtml",
      "nodeName": "HEAD",
      "numChildren": 5,
      "attrs": [],
      "isDisplayed": false
    },
    {
      "actor": "conn0.domnode42",
      "baseURI": "http://nicolagreco.com/test/",
      "parent": "conn0.domnode43",
      "nodeType": 1,
      "namespaceURI": "http://www.w3.org/1999/xhtml",
      "nodeName": "BODY",
      "numChildren": 1,
      "attrs": [],
      "isDisplayed": true
    }
  ],
  "from": "conn0.domwalker39"
}
```

It would be nice to trigger the action through the debugger without actually send javascript.

**Comment by nicola** [comment]
From the devtools on irc.mozilla.org

> I was looking at the code just now. There's a "property" request, but nothing like "setProperty".
> But I think there's a reason for that: if a property has a setter, or if the object is a proxy, then setting the property value could cause JavaScript to run. It could hit breakpoints,  throw an exception, or whatever.
> Hence, evaluation must be handled like other operations that continue the thread's execution, like eval. So at that point, the advantage over an actual 'eval' is less clear. But... I still think there should be a packet to just change an ordinary property's value.

**Comment by nicola** [comment]
I basically have written a plugin that
- asks what device and what tab you want to monitor
- inject that code mentioned in #1 using the idea in the comments below
- call reload when I press "Save" on a .css/.sass/etc file

#### advantages
- I can save a file and see live updates
- I avoid big hacks of mapping a file, setting a watcher
- I get updated css even if I compile them from sass or similar

#### however
- I cannot see live changes in the css on Firefox OS if I link to files locally (therefore in firefoxOS apps) - but can be fixed a) with the live-editing css, but is really what we want? b) injecting websockets c) injecting them via RDP and StyleEditor, but could be very painful

**Comment by nicola** [comment]
This was implemented in df67aaf142c96fcdb18c443d065f4e518ecfb066
